src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/InMemoryPartitionedAppTokenCacheAccessor.cs

```diff
@@ -215,14 +224,84 @@ public void SetiOSKeychainSecurityGroup(string keychainSecurityGroup)
+        }
+
+        internal class CacheWrapper
+        {
+            // perf: do not use ConcurrentDictionary.Values as it takes a lock
+            // internal for test only
+            internal readonly ConcurrentDictionary<string, ConcurrentDictionary<string, MsalAccessTokenCacheItem>> AccessTokenCacheDictionary;
+
+            private DateTimeOffset _lastExpirationScan;
+
+            private TimeSpan? _expirationScanFrequency;
+
+            internal CacheWrapper(TimeSpan? expirationScanFrequency)
+            {
+                _expirationScanFrequency = expirationScanFrequency;
+                _lastExpirationScan = DateTimeOffset.UtcNow;
+                AccessTokenCacheDictionary = new ConcurrentDictionary<string, ConcurrentDictionary<string, MsalAccessTokenCacheItem>>();
+            }
+
+            internal void UpdateFrequencyIfLower(TimeSpan expirationScanFrequency)
+            {
+                if (!_expirationScanFrequency.HasValue || _expirationScanFrequency.Value > expirationScanFrequency)
+                {
+                    _expirationScanFrequency = expirationScanFrequency;
+                }
+            }
+
+            // Called by multiple actions to see how long it's been since we last checked for expired items.
+            // If sufficient time has elapsed then a scan is initiated on a background task.
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            internal void StartScanForExpiredItemsIfNeeded()
+            {
+
+                if (_expirationScanFrequency.HasValue)
+                {
+                    var utcNow = DateTimeOffset.UtcNow;
+                    if (_expirationScanFrequency.Value < utcNow - _lastExpirationScan)
+                    {
+                        ScheduleTask(utcNow);
+                    }
+                }
+
+                void ScheduleTask(DateTimeOffset utcNow)
+                {
+                    _lastExpirationScan = utcNow;
+                    Task.Factory.StartNew(state => ScanForExpiredItems((CacheWrapper)state), this,
+                        CancellationToken.None, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default);
+                }
+            }
+
+            private static void ScanForExpiredItems(CacheWrapper cacheWrapper)
+            {
+                DateTimeOffset now = cacheWrapper._lastExpirationScan = DateTimeOffset.UtcNow;
+
+                foreach (KeyValuePair<string, ConcurrentDictionary<string, MsalAccessTokenCacheItem>> partitionItem in cacheWrapper.AccessTokenCacheDictionary)
+                {
+                    var partitionItemValue = partitionItem.Value;
+                    foreach (KeyValuePair<string, MsalAccessTokenCacheItem> item in partitionItemValue)
+                    {
+                        if (item.Value.ExpiresOn < now)
+                        {
+                            partitionItemValue.TryRemove(item.Key, out _);
```

Comment:
I'll let @pmaytak comment on this, but this implementation would not trigger an L2 cache update. 

We're thinking of adding proper L1 / L2 cache support in MSAL, with eviction options.